### ISSUE TITLE
monaco-yaml working with Angular 

### ISSUE THREAD

**Issue by fxck** [remcohaszing/monaco-yaml#252]
My goal is to get monaco-yaml working inside Angular app with monaco loaded from node_modules asynchronously. I've read all the related issues and tried countless things to get this working.

- I've managed to get everything working with `MonacoWebpackPlugin`, but this doubles my bundle size and build time.
- I tried using `@monaco-editor/loader`, this loads the editor, but when configured to [use the npm package](https://github.com/suren-atoyan/monaco-loader?tab=readme-ov-file#configure-the-loader-to-load-the-monaco-as-an-npm-package) it once again doubles the bundle size 
  -  on top of that, when `configureMonacoYaml` is ran, it throws `Uncaught Error: Unexpected usage` 
 - next I implemented the loading logic from [ngx-monaco-editor-v2](https://github.com/miki995/ngx-monaco-editor-v2/blob/master/projects/editor/src/lib/base-editor.ts#L80-L115), which loads Monaco from assets folder (where it is copied from node_modules)

The last method works fine, doesn't increase the bundle size, load monaco from node_modules, not from some remote CDN. But then comes the workers.

This is what I started with (when `onMonacoLoad` is called, `monaco` is loaded and globally available).

```typescript
onMonacoLoad: () => {

  window.MonacoEnvironment = {
    getWorker(_, label) {
      switch (label) {
        case 'editorWorkerService':
          return new Worker(new URL('/assets/monaco/monaco-editor/esm/vs/editor/editor.worker', import.meta.url));
        case 'json':
          return new Worker(new URL('/assets/monaco/monaco-editor/esm/vs/language/json/json.worker', import.meta.url));
        case 'yaml':
          return new Worker(new URL('monaco-yaml/yaml.worker', import.meta.url));
        default:
          throw new Error(`Unknown label ${label}`);
      }
    }
  };

  configureMonacoYaml(monaco, {
    enableSchemaRequest: true,
    hover: true,
    completion: true,
    validate: true,
    format: true,
    schemas: [
       // ...
    ]
  });

}
```

The first problem is, that no matter what path I put in `return new Worker(new URL('monaco-yaml/yaml.worker', import.meta.url));`, as soon as `configureMonacoYaml` it tries to load the worker from

<img width="1214" alt="image" src="https://github.com/user-attachments/assets/c3d281fa-d718-43ff-b3d3-980b1e1e407b">

This path seems to be based on `moduleId` from [here in the source](https://github.com/remcohaszing/monaco-yaml/blob/main/src/index.ts#L228). 

The first thing I tried was to use the same angular configuration that copies monaco from node_modules to assets, to copy `node_modules/monaco-yaml` to `/monaco-yaml`, this results in

> Uncaught SyntaxError: Cannot use import statement outside a module (at yaml.worker.js:2:1)

Next I tried to process that yaml.worker.js with webpack instead of just copying it, if the bundled version is placed where `configureMonacoYaml` expects it, it finally throws no errors... but the actual worker probably doesn't work, since yaml files where the schema worked just fine in the `MonacoWebpackPlugin` version now just say `Loading..` instead of suggesting.

I feel like I exhausted all my options. Couple of questions

- how does the `getWorker` function relate to where `configureMonacoYaml` actually looks for the worker?
- is `monaco-yml/yaml.worker.js` supposed to work as is, or is some additional compilation required?

Thanks!

**Comment by fxck** [comment]
Here are reproductions of different things I tried (`npm i`, `npm start`):

<br/>

### 1. Custom loader (ngx-monaco-editor-v2 style) + copying node_modules/monaco-yml/yaml.worker.js
- repository: https://github.com/fxck/angular-monaco-yaml/tree/custom-load-copy-from-modules
- relevant monaco-yaml code: https://github.com/fxck/angular-monaco-yaml/blob/custom-load-copy-from-modules/src/app/monaco/monaco.component.ts#L84-L115

Results in:
> Uncaught SyntaxError: Cannot use import statement outside a module (at yaml.worker.js:2:1)

<br/>

### 2. @monaco-editor/loader from CDN + copying node_modules/monaco-yml/yaml.worker.js
- repository: https://github.com/fxck/angular-monaco-yaml/tree/monaco-loader-cdn
- relevant monaco-yaml code: https://github.com/fxck/angular-monaco-yaml/blob/monaco-loader-cdn/src/app/monaco/monaco.component.ts#L83-L116

Results in:
> Uncaught SyntaxError: Cannot use import statement outside a module (at yaml.worker.js:2:1)

<br/>

### 3. @monaco-editor/loader from node_modules
- repository: https://github.com/fxck/angular-monaco-yaml/tree/monaco-loader-node-modules
- relevant monaco-yaml code: https://github.com/fxck/angular-monaco-yaml/blob/monaco-loader-node-modules/src/app/monaco/monaco.component.ts#L86-L119

Results in:
<img width="693" alt="image" src="https://github.com/user-attachments/assets/09908861-c726-4cf1-a443-e5f60d2d29a7">

<br/>

### 4. @monaco-editor/loader from CDN + processing yaml.worker.js with webpack (same with custom loader)
- repository: https://github.com/fxck/angular-monaco-yaml/tree/monaco-loader-cdn-webpack
- relevant monaco-yaml code: https://github.com/fxck/angular-monaco-yaml/blob/monaco-loader-cdn-webpack/src/app/monaco/monaco.component.ts#L83-L116
- custom webpack entry point: https://github.com/fxck/angular-monaco-yaml/blob/monaco-loader-cdn-webpack/src/webpack.config.ts#L6-L9

Results in no errors, but the yaml language server not working 
<img width="491" alt="image" src="https://github.com/user-attachments/assets/52c9a252-6c40-49fb-af0a-2c076eb66007">

**Comment by remcohaszing** [comment]
> I've managed to get everything working with `MonacoWebpackPlugin`, but this doubles my bundle size and build time.

I recommend going with this option.

All options will increase your users’ download sizes. Monaco editor is big. If you load your dependencies from a CDN, you just move the problem somewhere else so your tools won’t report it. It also means you now depend on the uptime and integrity of an external service.

> * how does the `getWorker` function relate to where `configureMonacoYaml` actually looks for the worker?

`window.MonacoEnvironment.getWorker` tells Monaco how to load workers. How to configure this, depends on your build toolchain. `configureMonacoYaml` loads a worker via Monaco editor.

Note that you can just define `window.MonacoEnvironment` at the top level of your module. You don’t need to configure this in some callback.

> * is `monaco-yml/yaml.worker.js` supposed to work as is, or is some additional compilation required?

It has dependencies and needs to be bundled.

---

I’m closing this as it’s a support question, not an actionable issue.

**Comment by fxck** [comment]
Thanks for the response @remcohaszing.

>  If you load your dependencies from a CDN ...  It also means you now depend on the uptime and integrity of an external service.
Yea, that's why I prefer when it loads it from assets/public folder (where you copy it from node_modules), which is what `ngx-monaco-editor-v2` essentially does, over what `@monaco-editor/loader` does by default (which is that CDN approach).

> `MonacoWebpackPlugin` ... I recommend going with this option.
The problem with this is that it doubles build time, build nodes resource requirements, as well as the initial load. If you use Monaco in some dialog component in a settings page, it's a big hit to take.

> Note that you can just define window.MonacoEnvironment at the top level of your module.
Yeah, I figured, but there didn't seem to be any difference. It get's called in both cases.

Anyway Monaco itself loads absolutely fine using either the `ngx-monaco-editor-v2` or `@monaco-editor/loader` method, the only problem seems to be getting the worker to run properly.

First thing that confuses me is (despite the `getWorker()` being called), that web workers defined here seems to be ignored, with this configuration

```typescript
window.MonacoEnvironment = {
  getWorker(_, label) {
    switch (label) {
      case 'editorWorkerService':
        return new Worker(new URL('/public/monaco/monaco-editor/esm/vs/editor/editor.worker', import.meta.url));
      case 'json':
        return new Worker(new URL('/public/monaco/monaco-editor/esm/vs/language/json/json.worker', import.meta.url));
      case 'yaml':
        return new Worker(new URL('/public/monaco-yaml/yaml.worker', import.meta.url));
      default:
        throw new Error(`Unknown label ${label}`);
    }
  }
};
```

It still loads jsonWorker from a different url.
<img width="715" alt="image" src="https://github.com/user-attachments/assets/72b9c716-ec84-4c1c-be4f-eb11a21a7492">

And it would try to load yaml.worker from different url as well, I actually had to change the `moduleId` here for it to start loading from the with place.

```typescript
  const workerManager = createWorkerManager(monaco, {
    label: "yaml",
    moduleId: "public/monaco-yaml/yaml.worker",
    createData
  });
```

So to me it seems like `getWorker()` is getting called but the worker it creates is completely ignored anyway and it spins up its own.

The other thing that confuses me is that Webpack, starting with version 5, [should compile web workers on its own](https://webpack.js.org/guides/web-workers/), so I shouldn't even need to manually add an entry point and make it compile the code. But it's clearly not happening. But even if I accept the game and place the manually compiled worker to the place it expects it, it still doesn't work, doesn't throw any error but just doesn't suggest anything (which is this case https://github.com/fxck/angular-monaco-yaml/tree/monaco-loader-cdn-webpack).
